src/desul/array.hpp

```diff
@@ -0,0 +1,310 @@
+      using pointer = value_type*;
+      using const_pointer = const value_type*;
+      using iterator = pointer;
+      using const_iterator = const_pointer;
+      using reverse_iterator = pointer;
+      using const_reverse_iterator = const_pointer;
+
+      constexpr reference at(size_type pos) {
+         if (pos >= size()) {
+            throw std::out_of_range{"desul::array::at detected out of range access"};
+         }
+
+         return elements[pos];
+      }
+
+      constexpr const_reference at(size_type pos) const {
+         if (pos >= size()) {
+            throw std::out_of_range{"desul::array::at detected out of range access"};
+         }
+
+         return elements[pos];
+      }
+
+      DESUL_HOST_DEVICE constexpr reference operator[](size_type pos) {
+         return elements[pos];
+      }
+
+      DESUL_HOST_DEVICE constexpr const_reference operator[](size_type pos) const {
+         return elements[pos];
+      }
+
+      DESUL_HOST_DEVICE constexpr reference front() {
+         return elements[0];
+      }
+
+      DESUL_HOST_DEVICE constexpr const_reference front() const {
+         return elements[0];
+      }
+
+      DESUL_HOST_DEVICE constexpr reference back() {
+         return elements[size() - 1];
+      }
+
+      DESUL_HOST_DEVICE constexpr const_reference back() const {
+         return elements[size() - 1];
+      }
+
+      DESUL_HOST_DEVICE constexpr pointer data() noexcept {
+         return elements;
+      }
+
+      DESUL_HOST_DEVICE constexpr const_pointer data() const noexcept {
+         return elements;
+      }
+
+      DESUL_HOST_DEVICE constexpr iterator begin() noexcept {
+         return elements;
+      }
+
+      DESUL_HOST_DEVICE constexpr const_iterator begin() const noexcept {
+         return elements;
+      }
+
+      DESUL_HOST_DEVICE constexpr const_iterator cbegin() const noexcept {
+         return elements;
+      }
+
+      DESUL_HOST_DEVICE constexpr iterator end() noexcept {
+         return &elements[size()];
+      }
+
+      DESUL_HOST_DEVICE constexpr const_iterator end() const noexcept {
+         return &elements[size()];
+      }
+
+      DESUL_HOST_DEVICE constexpr const_iterator cend() const noexcept {
+         return &elements[size()];
+      }
+
+      DESUL_HOST_DEVICE constexpr bool empty() const noexcept {
+         return size() == 0;
+      }
+
+      DESUL_HOST_DEVICE constexpr size_type size() const noexcept {
+         return size();
+      }
+
+      DESUL_HOST_DEVICE constexpr size_type max_size() const noexcept {
+         return size();
+      }
+
+      DESUL_HOST_DEVICE constexpr void fill(const T& value) {
+         for (std::size_t i = 0; i < size(); ++i) {
+            elements[i] = value;
+         }
+      }
+
+      DESUL_HOST_DEVICE constexpr void swap(array& other) noexcept(std::is_nothrow_swappable_v<T>) {
+         for (std::size_t i = 0; i < size(); ++i) {
+            swap(elements[i], other[i]);
```

Comment:
Add a `using std::swap`?

Reply:
I added a desul::swap function and am calling that directly now.